### PR TITLE
Dynamic function call

### PR DESCRIPTION
**Pull request by eeeebbbbrrrr**
(This replaces #1266 because I don't know how to use git.)

I was bored while on vacation, so I thought I'd start some code around being able to dynamically call any function that can be called from SQL.  

The implementation details are that this resolves the function by its name and argument types and calls it directly.  There's no Spi overhead here.  There is some allocation overhead and catalog access overhead, but that's insignificant when compared to everything Spi does.  (most) Any function in `pg_catalog.pg_proc` can be called.

We can bikeshed the API and names and such, but the general thought process is:

```sql
CREATE FUNCTION some_function(int4, text, bool[]) 
RETURNS text 
STRICT 
LANGUAGE sql AS $$ 
   SELECT '<whatever>' 
$$;
```

```rust
let result = pgrx::fcall::<&str>(
   "some_function",   // can be schema-qualified
   &[&Arg::Value(1i32), &Arg::Value("foo"), &Arg::Value(vec![true, true, false])]
);
assert_eq!(result, "<whatever>");
```

I'm happy for input now while it's still WIP, rather than later when it's mostly complete as I envision it.  We need this capability for PL/Rust so we need to make sure we get the user-facing API correct, at least.

Some things I don't think we should support in the first iteration of this:

- Postgres v11 (it EOL's in November -- not worth the extra code)
- Calling set-returning-functions (`RETURNS SET OF t` and `RETUNS TABLE (...)`)
- Functions with OUT parameters
- Automatically performing type casting/coersion for function arguments
- Named arguments

Some things I need to figure out:

- ~~Calling a function by Oid needs a collation Oid for any function that has TEXT arguments -- where does this come from?~~
- ~~I think `IntoDatum::into_datum(self)` should actually be `IntoDatum::as_datum(&self)` (or whatever).  The point is that converting a Rust value into a Datum doesn't need to consume the Rust value.  Not sure how this interacts with what @thomcc is working on~~ Doing a Clone is fine for now.

### CODE REVIEW

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (thomcc) [inline pgrx/src/pg_catalog/pg_proc.rs]: nit:
```suggestion
            v => panic!("unrecognized `ProVolatile`: `{:?}`", v as char),
```
since these use ASCII it will probably be easier to debug if it's formatted as a character than a number.

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (thomcc) [inline pgrx/src/pg_catalog/pg_proc.rs]:
```suggestion
            v => panic!("unrecognized `ProKind`: `{:?}`", v as char),
```

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (thomcc) [inline pgrx/src/pg_catalog/pg_proc.rs:47]: Should this be `#[non_exhaustive]`?

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (thomcc) [inline pgrx/src/pg_catalog/pg_proc.rs:22]: Should this be `#[non_exhaustive]`?

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (thomcc) [inline pgrx/src/pg_catalog/pg_proc.rs]: Are these actually safety constraints? If so, `get_attr` should be an `unsafe fn`. Otherwise, you should remove `SAFETY:` from the comment (here and below)

File: pgrx/src/fcall.rs
Comment (thomcc) [inline pgrx/src/fcall.rs]: Hmmmmmmmmmm...

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (thomcc) [inline pgrx/src/pg_catalog/pg_proc.rs]:
```suggestion
            v => panic!("unrecognized `ProArgMode`: `{:?}`", v as char),
```

**Review by thomcc — COMMENTED** [review]
rough pass, mostly skimming.

File: pgrx/src/fcall.rs
Comment (thomcc) [inline pgrx/src/fcall.rs]: Is it ever unsafe to use the wrong one? If not, I think we should offer a version that uses InvalidOid (which is the default), and one that allows the user to provide it.

That seems to match how function calls are performed in the PG sources -- usually with InvalidOid, but sometimes the collation is provided externally (I haven't threaded through to figure out what logic the higher level callers use to pass the collation in, but it seems like it varies).

File: pgrx/src/fcall.rs
Comment (thomcc) [inline pgrx/src/fcall.rs]: I wonder if it's better to catch these or allow them to propagate. I suppose it's more rusty to catch and return a result...

File: pgrx/src/pg_catalog/pg_proc.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/pg_catalog/pg_proc.rs]: Those notes were about the .unwrap() not panicking.  I'll change the wording.

File: pgrx/src/fcall.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: ???????

File: pgrx/src/fcall.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: > Is it ever unsafe to use the wrong one?

Using `InvalidOid` with a function that takes or returns a TEXT-type argument flat out fails with an error.  So we gotta provide some legitimate value somehow.  Jubilee had some ideas on Discord.

**Comment by eeeebbbbrrrr** [comment]
I think this is done.

We can bikeshed the "fcall" names a bit if anyone wants.  I was just looking for something easy to type and wouldn't overpower the function name string argument when reading code that uses this.

If this were javascript I'd have named this `$` so there's that.

**Comment by eeeebbbbrrrr** [comment]
This will ultimately need PR #1277.  So I'd actually like to get that merged first, then I'll do the work to update this.  The primary reason is not that the current PgList sucks (it does), but that it's hidden behind the `cshim` feature flag, and in order for `fcall` to work with PL/Rust it cannot be behind that feature.

File: pgrx/src/fcall.rs
Comment (thomcc) [inline pgrx/src/fcall.rs]: I literally came here to ask if we were doing this, and am super glad to see it's forbiddn!

File: pgrx/src/fcall.rs
Comment (thomcc) [inline pgrx/src/fcall.rs]: What to do is forbid it. It's entirely untyped. It's not entirely reasonable to use in a low level pgrx extension, but it should require unsafe code because you must know what you are doing.

File: pgrx/src/fcall.rs
Comment (thomcc) [inline pgrx/src/fcall.rs]: Just thought of this, but this is (potentially) an external call into a C postgres function, so if it errors it will longjmp, no? So I think we need a [`pg_guard_ffi_boundary`](https://github.com/pgcentralfoundation/pgrx/blob/develop/pgrx-pg-sys/src/submodules/ffi.rs#L92) around it, unless I'm mistaken

**Review by thomcc — CHANGES_REQUESTED** [review]
(Oops, forgot to mark as changes requested -- this shouldn't land without the error-handling guard. Might be nice to add a test calling some function which raises an error, but checking that it isn't longjmping over rust frames might be really annoying)

File: pgrx/src/fcall.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: I think it's more usable here to catch and convert to Result as the user may be implementing some kind of dynamic function calling system where they'd need to be able to recover from a function name that doesn't exist.

File: pgrx/src/fcall.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: yeah, great catch!  Didn't even occur to me.  :(

Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: Fixed up, with a test, in https://github.com/pgcentralfoundation/pgrx/pull/1279/commits/7616a6685c27f444afde64c6219b1ca2453c4a3d

**Comment by workingjubilee** [comment]
The List bugs that were blocking you should be fixed by a simple rebase on develop.

**Comment by eeeebbbbrrrr** [comment]
> The List bugs that were blocking you should be fixed by a simple rebase on develop.

yup, rebased this PR and CI passed and tests passed locally.  Excellent.

I just need an approval or two on this before we merge.

File: pgrx/src/fcall.rs
Comment (workingjubilee) [inline pgrx/src/fcall.rs]: Is it the case that default arguments must only come last? What about functions with mixed or named notation?

File: pgrx/src/fcall.rs
Comment (workingjubilee) [inline pgrx/src/fcall.rs]: What's the UX like for calling functions that normally used named notation? Is that even possible with this interface, or would we have to have another? Does Postgres internally canonize them to positional notation and you have to match the positional notation anyway, even if the argument fields get swapped?

Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: Yeah, postgres rewrites them into positional args.  I didn't feel like it was worth supporting.  We could in the future with a different function that takes a Map instead of a slice for the argument set.

File: pgrx/src/fcall.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/fcall.rs]: yes, default arguments must come last.

None of this supports named notation -- I should have listed that in the "things we don't support" up above.

File: pgrx/src/pg_catalog/mod.rs
Comment (workingjubilee) [inline pgrx/src/pg_catalog/mod.rs:1]: Hm.

**Review by workingjubilee — APPROVED** [review]
Looks good aiui with whatever naming stuff hashed out ~~and a comment about how functions using named notation may technically be "supported" but may be unstable (because they're secretly positional and there's not necessarily a promise on how exactly Postgres settles that positionality).~~

File: pgrx/src/pg_catalog/mod.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/pg_catalog/mod.rs:1]: > because they're secretly positional and there's not necessarily a promise on how exactly Postgres settles that positionality

There's a function that figures this out.  I ran across it last week but don't remember the name.

Keep in mind, we're not trying to emulate SQL here.  If a user wants all the niceties of SQL they can use SPI.

Comment (workingjubilee) [reply pgrx/src/pg_catalog/mod.rs:1]: agreed, I just asked to make sure we had an explanation in-hand if a user finds our behavior surprising.

**Comment by eeeebbbbrrrr** [comment]
Alrighty, @workingjubilee!  I believe this is done.

**Review by thomcc — APPROVED** [review]
My concerns are resolved, and it seems mostly fine.

File: pgrx/src/pg_catalog/mod.rs
Comment (eeeebbbbrrrr) [reply pgrx/src/pg_catalog/mod.rs:1]: Just a note to clarify that functions are not declared with named positional arguments, they're called that way.

Comment (workingjubilee) [reply pgrx/src/pg_catalog/mod.rs:1]: whoops!

**Pull request merged**
